### PR TITLE
Add generation pipeline for Image-to-Image LFM (#23)

### LINKED ISSUE #23: No Latent Flow Matching image-to-image generation pipeline
There is currently no dedicated pipeline for Image-to-Image generation. The current implementation lacks a proper folder structure, and documentation describing the workflow and intended approach.

The following changes should be made:
- Create a modular folder structure for the Image-to-Image pipeline.
- Isolate core pipeline steps into reusable functions/modules.
- Add documentation explaining the purpose and usage of the pipeline components.
- Add a CLI structure.
- Prepare the pipeline for future model training and inference integration.

### PR DESCRIPTION
**Pull request by MagicWiqqu**
## Description

This PR introduces a dedicated Image-to-Image pipeline structure. The previous implementation consisted primarily of notebook-based experimentation and lacked a modular architecture, reusable components, CLI support, and documentation. The new structure separates core pipeline functionality into reusable modules, establishes a clear project organization, and provides documentation describing the workflow and intended usage. The pipeline is also prepared for future integration of model training and inference components.

## Related Issue

Closes #23 

## Proposed Changes

* Creates a modular folder structure for the Image-to-Image pipeline.
* Refactors pipeline logic into reusable functions and modules.
* Adds documentation describing the pipeline architecture, workflow, and usage.
* Introduces a CLI entry point for pipeline execution.
* Establishes a foundation for future model training and inference integration.

### CODE REVIEW

File: src/pipeline/inference/sliding_window_inference.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+        raise ValueError("`chunk_size` must be greater than 0")
+
+    if halo_size < 0:
+        raise ValueError("`halo_size` must be non-negative")
+
+    if image_size <= 0:
+        raise ValueError("`image_size` must be greater than 0")
+
+    if factor <= 0:
+        raise ValueError("`factor` must be greater than 0")
```
Comment (Couch-bit) [inline src/pipeline/inference/sliding_window_inference.py]: Generally, the validation in the functions are quite inconsistent. Sometimes you check the type and sometimes you don't - I would lean towards not checking since it's a ML pipeline which is run locally. Also sometimes the data validation has varying degrees of complexity across functions (this is mainly a problem for metrics).

File: src/pipeline/config.py [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+
+    # Halos allow the sliding window to be lossless.
+    halo_encoder: int = 8
+    halo_decoder: int = 2
+
+    # MAISI VAE spatial scaling factor.
+    # Encoder: image space -> latent space, /4
+    # Decoder: latent space -> image space, *4
+    encoder_factor: int = 4
+    decoder_factor: int = 4
```
Comment (Couch-bit) [inline src/pipeline/config.py]: These are always identical, so they should be one parameter.

File: src/pipeline/config.py [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+    weights_dir: Path = Path("src/pipeline/weights")
+
+    vae_weight_path: Path = Path("src/pipeline/weights/autoencoder_v1.pt")
+    rflow_weight_path: Path = Path("src/pipeline/weights/diff_unet_3d_rflow-ct.pt")
+
+    # Inference config
+    cts_per_patient: int = 1
+    steps: int = 30
+    latent_scale: float = 1.0
+    device: str = "cuda"
```
Comment (Couch-bit) [inline src/pipeline/config.py]: Whether to use cuda or cpu should be automatically detected I think.

File: src/pipeline/config.py [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+    latent_ct_dir: Path = Path("RESULTS/MAISI_TESTING/latents/test")
+    generated_ct_dir: Path = Path("RESULTS/MAISI_TESTING/generated_ct/test")
+    metrics_dir: Path = Path("RESULTS/MAISI_TESTING/metrics")
+    logs_dir: Path = Path("RESULTS/MAISI_TESTING/logs")
+
+    # Model weights
+    weights_dir: Path = Path("src/pipeline/weights")
+
+    vae_weight_path: Path = Path("src/pipeline/weights/autoencoder_v1.pt")
+    rflow_weight_path: Path = Path("src/pipeline/weights/diff_unet_3d_rflow-ct.pt")
```
Comment (Couch-bit) [inline src/pipeline/config.py]: Why explicitly store the root directories if they are not used?

File: pyproject.toml [hunk 5]
```diff
@@ -6,12 +6,14 @@ requires-python = "==3.11.15"
...
     "monai>=1.5.2,<2.0.0",
     "mypy>=1.20.0,<2.0.0",
     "nibabel>=5.4.2,<6.0.0",
-    "numpy>=2.4.4,<3.0.0",
+    "numpy==2.4.4",
     "optuna>=4.8.0,<5.0.0",
+    "pandas>=2.2",
     "ruff>=0.15.9,<1.0.0",
     "scikit-image>=0.26.0,<1.0.0",
     "simpleitk>=2.5.3,<3.0.0",
```
Comment (Couch-bit) [inline pyproject.toml]: All added packages should have an upper bound equal to the next major version (or less if there is a justification).

File: pyproject.toml [hunk 6]
```diff
@@ -74,6 +76,10 @@ quote-style = "double"
 python_version = "3.11"
 strict = true
 
+[[tool.mypy.overrides]]
+module = ["lpips"]
+ignore_missing_imports = true
+
```
Comment (Couch-bit) [inline pyproject.toml]: This is redundant since lpips is in the requirements. I would just assume the import is mandatory both here and in the code.

File: pyproject.toml [hunk 7]
```diff
@@ -82,3 +88,10 @@ exclude = [
 
 # Standard stability settings
 warn_unused_configs = true
+
+[dependency-groups]
+dev = [
+    "pandas-stubs>=3.0.3.260530",
+    "scipy-stubs>=1.17.1.5",
+    "types-tqdm>=4.67.3.20260518",
+]
```
Comment (Couch-bit) [inline pyproject.toml:93]: This is ok, but will need an update to the docs in the future as they currently don't mention dev.

File: src/pipeline/inference/sliding_window_inference.py [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+
+def sliding_window_inference(
+    image: torch.Tensor,
+    chunk_size: int,
+    halo_size: int,
+    image_size: int,
+    model: Callable[[torch.Tensor], torch.Tensor],
+    model_type: str,
+    factor: int,
+) -> torch.Tensor:
```
Comment (Couch-bit) [inline src/pipeline/inference/sliding_window_inference.py]: This function is also used by the trianing pipeline, should be in the common directory along with `data_utils.py`.

File: src/pipeline/inference/run_generation.py [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+                "Rectified flow checkpoint does not match the current model "
+                "architecture.\n"
+                f"Layer: {first_layer_key}\n"
+                f"Checkpoint shape: {checkpoint_shape}\n"
+                f"Model shape: {model_shape}\n\n"
+                "This probably means that the checkpoint was not trained or "
+                "adapted for the current `rflow_config`. If your model expects "
+                "8 input channels, the checkpoint must also contain 8-channel "
+                "weights"
+            )
```
Comment (Couch-bit) [inline src/pipeline/inference/run_generation.py]: This logic is redundant, you should assume they are in the correct format.

File: src/pipeline/inference/run_generation.py [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+    )
+
+    dataset = PatientConditionDataset(config.latent_ct_dir)
+
+    loader = ThreadDataLoader(
+        cast(Any, dataset),
+        batch_size=1,
+        shuffle=False,
+        pin_memory=device.type == "cuda",
+    )
```
Comment (Couch-bit) [inline src/pipeline/inference/run_generation.py:284]: Why the cast to Any?

**Review by Couch-bit — COMMENTED** [review]
There are no `__init__.py` files in nested directories (data, evaluation, inference) - they should also be there.

File: src/pipeline/config.py (hunk 2 above)
Comment (MagicWiqqu) [reply src/pipeline/config.py]: OK, I've merged it into one parameter `encoder_decoder_factor`.

File: src/pipeline/inference/sliding_window_inference.py (hunk 8 above)
Comment (MagicWiqqu) [reply src/pipeline/inference/sliding_window_inference.py]: OK, I've moved it to `data_utils.py`.

**Comment by MagicWiqqu** [comment]
> There are no `__init__.py` files in nested directories (data, evaluation, inference) - they should also be there.

For some reason I cannot answer directly to the comment, but I've added them now.

File: src/pipeline/inference/sliding_window_inference.py (hunk 1 above)
Comment (MagicWiqqu) [reply src/pipeline/inference/sliding_window_inference.py]: Do you wish for not checking at all, or just not checking types? I assume to drop checking types, but I want to have it clarified.

File: src/pipeline/config.py (hunk 3 above)
Comment (MagicWiqqu) [reply src/pipeline/config.py]: OK, I've changed it to auto-detect now.

File: pyproject.toml (hunk 5 above)
Comment (MagicWiqqu) [reply pyproject.toml]: OK, I've added upper bounds.

File: pyproject.toml (hunk 6 above)
Comment (MagicWiqqu) [reply pyproject.toml]: OK, added # type: ignore instead to force mypy to calm down.

File: src/pipeline/config.py (hunk 4 above)
Comment (MagicWiqqu) [reply src/pipeline/config.py]: It's a leftover from the previous overhaul. Will fix it now.

File: src/pipeline/inference/run_generation.py (hunk 10 above)
Comment (MagicWiqqu) [reply src/pipeline/inference/run_generation.py:284]: It was a temporary fix to stop mypy from throwing errors. Did a proper cast now to solve it.

File: src/pipeline/inference/run_generation.py (hunk 9 above)
Comment (MagicWiqqu) [reply src/pipeline/inference/run_generation.py]: OK, I shortened it now.

File: pyproject.toml (hunk 7 above)
Comment (MagicWiqqu) [reply pyproject.toml:93]: Yeah, it is needed for mypy to be a bit more friendly.

File: src/pipeline/inference/sliding_window_inference.py (hunk 1 above)
Comment (Couch-bit) [reply src/pipeline/inference/sliding_window_inference.py]: Yes, just drop checking types, value checking is useful.

**Review by Couch-bit — COMMENTED** [review]
This looks okay now. Is there anything else this PR should add before merging?

**Comment by MagicWiqqu** [comment]
> This looks okay now. Is there anything else this PR should add before merging?

When I fix the issues, it should be ready to merge. I will change it to 'ready for review' soon.

File: src/pipeline/inference/run_generation.py (hunk 10 above)
Comment (Couch-bit) [reply src/pipeline/inference/run_generation.py:284]: You didn't seem to change anything.

Comment (MagicWiqqu) [reply src/pipeline/inference/run_generation.py:284]: Now it is `dataset=cast(MonaiDataset, dataset),`, otherwise mypy won't stop throwing errors at it. Would you prefer other solution?

Comment (Couch-bit) [reply src/pipeline/inference/run_generation.py:284]: I mean that you didn't push this change.

File: src/pipeline/inference/sliding_window_inference.py (hunk 1 above)
Comment (MagicWiqqu) [reply src/pipeline/inference/sliding_window_inference.py]: OK, dropped checking types.

File: src/pipeline/inference/run_generation.py (hunk 10 above)
Comment (MagicWiqqu) [reply src/pipeline/inference/run_generation.py:284]: It has been pushed via commit 25b98a8. I pushed again with another commit. May you check? From my side, I see `dataset=cast(MonaiDataset, dataset)`.

File: src/pipeline/evaluation/metrics.py [hunk 11]
```diff
@@ -0,0 +1,1072 @@
...
+
+    ref_tensor = _center_slices_for_lpips(
+        ref,
+        max_slices=max_slices,
+    ).to(device)
+
+    with torch.no_grad():
+        score = lpips_model(pred_tensor, ref_tensor)
+
+    return float(score.mean().item())
```
Comment (Couch-bit) [inline src/pipeline/evaluation/metrics.py]: This uses only one slice which means the metric is very noisy. You should either calculate it for all or a large proportion of evenly spaced slices. Another option is to calculate normal perceptual similarity instead using 3D or 2.5D calculations, which may actually be more appropriate for scientific images.

**Review by Couch-bit — APPROVED** [review]
LGTM
